File: apps/android/core/src/main/kotlin/com/wordocious/core/Hub.kt
package com.wordocious.core

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Hubbub — seven-letter hub game (More Games §12). 1:1 port of
 * packages/core/src/games/hub.ts; pinned by hub-fixtures.json (HubFixtureTest).
 * Words of 4+ letters using only the seven letters and containing the center;
 * 4 letters = 1 point, else length, pangram +7; bonus words score 0. Ranks by
 * integer maths (points*100 >= pct*max); Hubbub (50%) = solved. Play continues
 * after the win; End finalizes a loss when below Hubbub. Event sigils + = ? ! #
 */
const val HUB_DAILY_EPOCH = "2026-09-23"
const val HUB_LETTERS = 7
const val HUB_MIN_WORD = 4
const val HUB_TOTAL_BOARDS = 20
const val HUB_SOLVED_RANK = 6
val HUB_RANKS: List<Pair<String, Int>> = listOf(
    "Hush" to 0, "Murmur" to 5, "Chatter" to 12, "Banter" to 20, "Clamor" to 30, "Racket" to 40,
    "Hubbub" to 50, "Uproar" to 70, "Thunder" to 85, "Pandemonium" to 100,
)

@Serializable
data class HubPuzzle(val id: String, val letters: String, val words: List<String>, val bonus: List<String>, val pangrams: List<String>, val max: Int)

@Serializable
data class HubBank(val version: Int, val epoch: String, val daily: List<HubPuzzle>, val extra: List<HubPuzzle>) {
    companion object {
        private val json = Json { ignoreUnknownKeys = true }
        fun parse(text: String): HubBank? = runCatching { json.decodeFromString<HubBank>(text) }.getOrNull()
        /** The bundled bank (core resources/data/hub-puzzles.json, sha-guarded to match the web copy). */
        val bundled: HubBank? by lazy {
            HubBank::class.java.classLoader?.getResourceAsStream("data/hub-puzzles.json")?.bufferedReader()?.use { it.readText() }?.let { parse(it) }
        }
    }
}

fun hubPuzzleForDay(bank: HubBank, day: String): HubPuzzle? = if (bank.daily.isEmpty()) null else bank.daily[Bank.indexForDay(day, bank.daily.size, bank.epoch)]
fun hubPuzzleForSeed(bank: HubBank, seed: String): HubPuzzle? {
    val pool = if (bank.extra.isEmpty()) bank.daily else bank.extra
    return if (pool.isEmpty()) null else pool[Bank.indexForSeed(seed, pool.size)]
}
fun hubDailyNumber(day: String): Int { val idx = Bank.dayIndex(day, HUB_DAILY_EPOCH) ?: return 1; return maxOf(1, idx + 1) }

// ── Scoring ──────────────────────────────────────────────────────────────

fun hubIsPangram(word: String, letters: String): Boolean = letters.all { it in word }
fun hubWordScore(word: String, letters: String): Int = (if (word.length == 4) 1 else word.length) + (if (hubIsPangram(word, letters)) 7 else 0)
fun hubRankIndex(points: Int, max: Int): Int {
    if (max <= 0) return 0
    var i = 0
    for (k in HUB_RANKS.indices) if (points * 100 >= HUB_RANKS[k].second * max) i = k
    return i
}
fun hubRankThreshold(rank: Int, max: Int): Int = ((HUB_RANKS[rank].second * max) + 99) / 100
fun hubGuessCount(rankIndex: Int): Int = HUB_RANKS.size - rankIndex
fun hubBoardsSolved(points: Int, max: Int): Int = if (max <= 0) 0 else minOf(HUB_TOTAL_BOARDS, (points * HUB_TOTAL_BOARDS) / max)
fun hubStartsWithToken(word: String): String = "${word.take(2)}${word.length}"

// ── Reducer ──────────────────────────────────────────────────────────────

enum class HubStatus(val key: String) { PLAYING("playing"), WON("won"), LOST("lost") }
enum class HubReject(val key: String) { ENDED("ended"), SHORT("short"), CENTRE("centre"), LETTERS("letters"), FOUND("found"), NOTWORD("notword") }

data class HubState(
    val seed: String, val id: String, val letters: String, val words: List<String>, val bonus: List<String>, val pangrams: List<String>, val max: Int,
    val found: List<String>, val bonusFound: List<String>, val revealed: List<String>, val hinted: List<String>,
    val points: Int, val hintsUsed: Int, val events: List<String>, val status: HubStatus, val ended: Boolean, val reject: HubReject?,
    val startTime: Long, val endTime: Long?,
) {
    val centre: Char get() = letters[0]
    val rank: Int get() = hubRankIndex(points, max)
    val rankName: String get() = HUB_RANKS[rank].first
    val guessCount: Int get() = hubGuessCount(rank)
    val boardsSolved: Int get() = hubBoardsSolved(points, max)
    fun nextUnfound(skipHinted: Boolean = false): String? = words.firstOrNull { it !in found && (!skipHinted || it !in hinted) }

    companion object {
        fun create(p: HubPuzzle, seed: String, startTime: Long) = HubState(
            seed, p.id, p.letters, p.words, p.bonus, p.pangrams, p.max, emptyList(), emptyList(), emptyList(), emptyList(),
            0, 0, emptyList(), HubStatus.PLAYING, false, null, startTime, null,
        )
    }
}

sealed class HubAction {
    data class Submit(val word: String) : HubAction()
    object HintStart : HubAction()
    object HintReveal : HubAction()
    object End : HubAction()
    object Finish : HubAction()
}

private fun settle(s: HubState, now: Long): HubState =
    if (s.status == HubStatus.PLAYING && s.rank >= HUB_SOLVED_RANK) s.copy(status = HubStatus.WON, endTime = now) else s

fun hubReduce(s: HubState, a: HubAction, now: Long = 0): HubState {
    if (a is HubAction.Finish) return if (s.status == HubStatus.PLAYING) s else s.copy(endTime = s.endTime ?: now)
    if (s.ended) return if (a is HubAction.Submit) s.copy(reject = HubReject.ENDED) else s
    return when (a) {
        is HubAction.Submit -> {
            val word = a.word.uppercase()
            when {
                word.length < HUB_MIN_WORD -> s.copy(reject = HubReject.SHORT)
                s.centre !in word -> s.copy(reject = HubReject.CENTRE)
                word.any { it !in s.letters } -> s.copy(reject = HubReject.LETTERS)
                word in s.found || word in s.bonusFound -> s.copy(reject = HubReject.FOUND)
                word in s.words -> settle(s.copy(found = s.found + word, points = s.points + hubWordScore(word, s.letters), events = s.events + "+$word", reject = null), now)
                word in s.bonus -> s.copy(bonusFound = s.bonusFound + word, events = s.events + "=$word", reject = null)
                else -> s.copy(reject = HubReject.NOTWORD)
            }
        }
        is HubAction.HintStart -> {
            val target = s.nextUnfound(skipHinted = true) ?: return s.copy(reject = null)
            s.copy(hinted = s.hinted + target, hintsUsed = s.hintsUsed + 1, events = s.events + "?${hubStartsWithToken(target)}", reject = null)
        }
        is HubAction.HintReveal -> {
            val target = s.nextUnfound() ?: return s.copy(reject = null)
            settle(s.copy(found = s.found + target, revealed = s.revealed + target, points = s.points + hubWordScore(target, s.letters), hintsUsed = s.hintsUsed + 2, events = s.events + "!$target", reject = null), now)
        }
        is HubAction.End -> {
            val n = s.copy(ended = true, events = s.events + "#", reject = null)
            if (n.status == HubStatus.PLAYING) n.copy(status = HubStatus.LOST, endTime = now) else n
        }
        is HubAction.Finish -> s
    }
}

// ── Matches row ↔ state ─────────────────────────────────────────────────

fun hubMatchRow(s: HubState): Pair<List<String>, List<String>> =
    listOf(s.id, s.letters, "${s.max}", "${s.words.size}", "${s.pangrams.size}") to s.events

data class HubReconstruction(
    val id: String, val letters: String, val max: Int, val wordCount: Int, val pangramCount: Int,
    val found: List<String>, val bonusFound: List<String>, val revealed: List<String>, val hints: List<String>,
    val points: Int, val hintsUsed: Int, val rank: Int, val rankName: String, val ended: Boolean, val solved: Boolean,
)

fun reconstructHub(solutions: List<String>, guesses: List<String>): HubReconstruction? {
    if (solutions.size < 3) return null
    val letters = solutions[1]
    if (letters.length != 7 || !letters.all { it in 'A'..'Z' }) return null
    val max = solutions[2].toIntOrNull() ?: return null
    if (max <= 0) return null
    val found = ArrayList<String>(); val bonusFound = ArrayList<String>(); val revealed = ArrayList<String>(); val hints = ArrayList<String>()
    var points = 0; var hintsUsed = 0; var ended = false
    fun isWord(w: String) = w.length >= 4 && w.all { it in 'A'..'Z' }
    for (ev in guesses) {
        if (ev == "#") { ended = true; continue }
        val sigil = ev.firstOrNull() ?: continue
        val rest = ev.substring(1)
        when {
            (sigil == '+' || sigil == '!') && isWord(rest) && rest !in found -> { found.add(rest); points += hubWordScore(rest, letters); if (sigil == '!') { revealed.add(rest); hintsUsed += 2 } }
            sigil == '=' && isWord(rest) -> bonusFound.add(rest)
            sigil == '?' -> { hints.add(rest); hintsUsed += 1 }
        }
    }
    val rank = hubRankIndex(points, max)
    return HubReconstruction(solutions[0], letters, max, solutions.getOrNull(3)?.toIntOrNull() ?: 0, solutions.getOrNull(4)?.toIntOrNull() ?: 0,
        found, bonusFound, revealed, hints, points, hintsUsed, rank, HUB_RANKS[rank].first, ended, rank >= HUB_SOLVED_RANK)
}

File: apps/ios/Sources/Core/Hub.swift
import Foundation

// Hubbub — seven-letter hub game (More Games §12). 1:1 port of
// packages/core/src/games/hub.ts; pinned by hub-fixtures.json (HubFixtureTests).
// Words of 4+ letters using only the seven letters and containing the centre;
// 4 letters = 1 point, else length, pangram +7; EVERY accepted word scores (founder 2026-09-25), max = core list. Ranks by
// integer maths (points*100 >= pct*max); Hubbub (50%) = solved. Play continues
// after the win; End finalises a loss when below Hubbub. Event sigils + = ? ! #

public let HUB_DAILY_EPOCH = "2026-09-23"
public let HUB_LETTERS = 7
public let HUB_MIN_WORD = 4
public let HUB_TOTAL_BOARDS = 20
public let HUB_SOLVED_RANK = 6
public let HUB_RANKS: [(name: String, pct: Int)] = [
    ("Hush", 0), ("Murmur", 5), ("Chatter", 12), ("Banter", 20), ("Clamor", 30), ("Racket", 40),
    ("Hubbub", 50), ("Uproar", 70), ("Thunder", 85), ("Pandemonium", 100),
]

public struct HubPuzzle: Codable, Equatable {
    public let id: String
    public let letters: String
    public let words: [String]
    public let bonus: [String]
    public let pangrams: [String]
    public let max: Int
    public init(id: String, letters: String, words: [String], bonus: [String], pangrams: [String], max: Int) {
        self.id = id; self.letters = letters; self.words = words; self.bonus = bonus; self.pangrams = pangrams; self.max = max
    }
}

public struct HubBank: Codable {
    public let version: Int
    public let epoch: String
    public let daily: [HubPuzzle]
    public let extra: [HubPuzzle]
    public init(version: Int, epoch: String, daily: [HubPuzzle], extra: [HubPuzzle]) { self.version = version; self.epoch = epoch; self.daily = daily; self.extra = extra }
    public static func load(from data: Data) -> HubBank? { try? JSONDecoder().decode(HubBank.self, from: data) }
}

public func hubPuzzleForDay(_ bank: HubBank, day: String) -> HubPuzzle? {
    guard !bank.daily.isEmpty else { return nil }
    return bank.daily[Bank.indexForDay(day, n: bank.daily.count, epoch: bank.epoch)]
}
public func hubPuzzleForSeed(_ bank: HubBank, seed: String) -> HubPuzzle? {
    let pool = bank.extra.isEmpty ? bank.daily : bank.extra
    guard !pool.isEmpty else { return nil }
    return pool[Bank.indexForSeed(seed, n: pool.count)]
}
public func hubDailyNumber(_ day: String) -> Int {
    guard let idx = Bank.dayIndex(day, epoch: HUB_DAILY_EPOCH) else { return 1 }
    return max(1, idx + 1)
}

// MARK: - Scoring

public func hubIsPangram(_ word: String, letters: String) -> Bool { letters.allSatisfy { word.contains($0) } }
public func hubWordScore(_ word: String, letters: String) -> Int { (word.count == 4 ? 1 : word.count) + (hubIsPangram(word, letters: letters) ? 7 : 0) }
public func hubRankIndex(points: Int, max: Int) -> Int {
    guard max > 0 else { return 0 }
    var i = 0
    for k in 0..<HUB_RANKS.count where points * 100 >= HUB_RANKS[k].pct * max { i = k }
    return i
}
public func hubRankThreshold(_ rank: Int, max: Int) -> Int { Int((Double(HUB_RANKS[rank].pct * max) / 100).rounded(.up)) }
public func hubGuessCount(_ rankIndex: Int) -> Int { HUB_RANKS.count - rankIndex }
public func hubBoardsSolved(points: Int, max: Int) -> Int { max > 0 ? min(HUB_TOTAL_BOARDS, (points * HUB_TOTAL_BOARDS) / max) : 0 }
public func hubStartsWithToken(_ word: String) -> String { "\(word.prefix(2))\(word.count)" }

// MARK: - Reducer

public enum HubStatus: String, Codable { case playing, won, lost }
public enum HubReject: String, Codable { case ended, short, centre, letters, found, notword }

public struct HubState: Codable, Equatable {
    public let seed: String
    public let id: String
    public let letters: String
    public let words: [String]
    public let bonus: [String]
    public let pangrams: [String]
    public let max: Int
    public var found: [String]
    public var bonusFound: [String]
    public var revealed: [String]
    public var hinted: [String]
    public var points: Int
    public var hintsUsed: Int
    public var events: [String]
    public var status: HubStatus
    public var ended: Bool
    public var reject: HubReject?
    public var startTime: Double
    public var endTime: Double?

    public init(puzzle p: HubPuzzle, seed: String, startTime: Double) {
        self.seed = seed; id = p.id; letters = p.letters; words = p.words; bonus = p.bonus; pangrams = p.pangrams; max = p.max
        found = []; bonusFound = []; revealed = []; hinted = []; points = 0; hintsUsed = 0; events = []
        status = .playing; ended = false; reject = nil; self.startTime = startTime; endTime = nil
    }

    public var centre: Character { letters.first! }
    public var rank: Int { hubRankIndex(points: points, max: max) }
    public var rankName: String { HUB_RANKS[rank].name }
    public var guessCount: Int { hubGuessCount(rank) }
    public var boardsSolved: Int { hubBoardsSolved(points: points, max: max) }
    public func nextUnfound(skipHinted: Bool = false) -> String? { words.first { !found.contains($0) && (!skipHinted || !hinted.contains($0)) } }
}

public enum HubAction: Equatable {
    case submit(String)
    case hintStart
    case hintReveal
    case end
    case finish
}

private func settle(_ s: HubState, _ now: Double) -> HubState {
    if s.status == .playing && s.rank >= HUB_SOLVED_RANK { var n = s; n.status = .won; n.endTime = now; return n }
    return s
}

public func hubReduce(_ s: HubState, _ a: HubAction, now: Double = 0) -> HubState {
    if case .finish = a { if s.status == .playing { return s }; var n = s; n.endTime = s.endTime ?? now; return n }
    if s.ended { if case .submit = a { var n = s; n.reject = .ended; return n }; return s }
    switch a {
    case .submit(let raw):
        let word = raw.uppercased()
        var n = s
        if word.count < HUB_MIN_WORD { n.reject = .short; return n }
        if !word.contains(s.centre) { n.reject = .centre; return n }
        for ch in word where !s.letters.contains(ch) { n.reject = .letters; return n }
        if s.found.contains(word) || s.bonusFound.contains(word) { n.reject = .found; return n }
        if s.words.contains(word) {
            n.found.append(word); n.points += hubWordScore(word, letters: s.letters); n.events.append("+\(word)"); n.reject = nil
            return settle(n, now)
        }
        // Founder (2026-09-25): every accepted word scores; "=" keeps marking the rarer ones.
        if s.bonus.contains(word) {
            n.bonusFound.append(word); n.points += hubWordScore(word, letters: s.letters); n.events.append("=\(word)"); n.reject = nil
            return settle(n, now)
        }
        n.reject = .notword
        return n
    case .hintStart:
        var n = s; n.reject = nil
        guard let target = s.nextUnfound(skipHinted: true) else { return n }
        n.hinted.append(target); n.hintsUsed += 1; n.events.append("?\(hubStartsWithToken(target))")
        return n
    case .hintReveal:
        var n = s; n.reject = nil
        guard let target = s.nextUnfound() else { return n }
        n.found.append(target); n.revealed.append(target); n.points += hubWordScore(target, letters: s.letters); n.hintsUsed += 2; n.events.append("!\(target)")
        return settle(n, now)
    case .end:
        var n = s; n.ended = true; n.events.append("#"); n.reject = nil
        if n.status == .playing { n.status = .lost; n.endTime = now }
        return n
    case .finish:
        return s
    }
}

// MARK: - Matches row ↔ state

public func hubMatchRow(_ s: HubState) -> (solutions: [String], guesses: [String]) {
    ([s.id, s.letters, "\(s.max)", "\(s.words.count)", "\(s.pangrams.count)"], s.events)
}

public struct HubReconstruction: Equatable {
    public let id: String, letters: String, max: Int, wordCount: Int, pangramCount: Int
    public let found: [String], bonusFound: [String], revealed: [String], hints: [String]
    public let points: Int, hintsUsed: Int, rank: Int, rankName: String, ended: Bool, solved: Bool
}

public func reconstructHub(solutions: [String], guesses: [String]) -> HubReconstruction? {
    guard solutions.count >= 3 else { return nil }
    let letters = solutions[1]
    guard letters.count == 7, letters.allSatisfy({ ("A"..."Z").contains($0) }), let max = Int(solutions[2]), max > 0 else { return nil }
    var found: [String] = [], bonusFound: [String] = [], revealed: [String] = [], hints: [String] = []
    var points = 0, hintsUsed = 0, ended = false
    let isWord = { (w: String) in w.count >= 4 && w.allSatisfy { ("A"..."Z").contains($0) } }
    for ev in guesses {
        if ev == "#" { ended = true; continue }
        guard let sigil = ev.first else { continue }
        let rest = String(ev.dropFirst())
        if (sigil == "+" || sigil == "!") && isWord(rest) && !found.contains(rest) {
            found.append(rest); points += hubWordScore(rest, letters: letters)
            if sigil == "!" { revealed.append(rest); hintsUsed += 2 }
        } else if sigil == "=" && isWord(rest) && !bonusFound.contains(rest) { bonusFound.append(rest); points += hubWordScore(rest, letters: letters) }
        else if sigil == "?" { hints.append(rest); hintsUsed += 1 }
    }
    let rank = hubRankIndex(points: points, max: max)
    return HubReconstruction(id: solutions[0], letters: letters, max: max, wordCount: solutions.count > 3 ? Int(solutions[3]) ?? 0 : 0,
                             pangramCount: solutions.count > 4 ? Int(solutions[4]) ?? 0 : 0, found: found, bonusFound: bonusFound,
                             revealed: revealed, hints: hints, points: points, hintsUsed: hintsUsed, rank: rank,
                             rankName: HUB_RANKS[rank].name, ended: ended, solved: rank >= HUB_SOLVED_RANK)
}
